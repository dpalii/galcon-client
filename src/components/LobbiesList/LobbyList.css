.lobbies {
    border: 1px solid var(--primary-color);
    border-collapse: collapse;
    margin-bottom: 10px;
}

.lobbies td,
.lobbies th {
    border: 1px solid var(--primary-color);
    padding: 3px;
}

.pagination {
    display: flex;
    flex-direction: row;
    justify-content: center;
    margin-bottom: 10px;
}

.lobbies-wrapper {
    height: 200px;
}

.page {
    padding: 5px 10px;
}