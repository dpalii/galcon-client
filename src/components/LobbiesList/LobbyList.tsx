import { useContext, useEffect, useState } from 'react'
import { SocketContext } from '../../contexts/SocketContext'
import './LobbyList.css'
import { IncomingEvents, Lobby, OutgoingEvents } from '../../types'

export interface LobbyListProps {
    joinLobby: (gameId: string) => void
    close: () => void
}

export function LobbyList({ joinLobby, close }: LobbyListProps) {
    const PAGE_SIZE = 5
    const socket = useContext(SocketContext)
    const [lobbies, setLobbies] = useState<Lobby[]>([])
    const [page, setPage] = useState(0)

    const getLobbies = () => {
        socket.emit(OutgoingEvents.GET_LOBBY_LIST, (lobbyList: Lobby[]) => {
            setLobbies(lobbyList)
        })
    }
    
    const getLastPage = () => {
        return Math.floor(lobbies.length / PAGE_SIZE)
    }

    useEffect(getLobbies, [socket])

    useEffect(() => {
        socket.on(IncomingEvents.GAME_CREATED, getLobbies)

        return () => {
            socket.off(IncomingEvents.GAME_CREATED, getLobbies)
        }
    })

    return (
        <div className="backdrop" onClick={() => close()}>
            <div className="modal" onClick={(e) => e.stopPropagation()}>
                <h2 className="heading">LOBBIES:</h2>
                <div className='lobbies-wrapper'>
                    <table className='lobbies'>
                        <thead>
                            <tr>
                                <th>ID</th>
                                <th>User Amount</th>
                                <th></th>
                            </tr>
                        </thead>
                        <tbody>
                            {
                                lobbies.slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE).map((lobby) => (
                                    <tr>
                                        <td>{lobby.id}</td>
                                        <td>{lobby.usersAmount}</td>
                                        <td>
                                            <button className='btn btn-sm' onClick={() => joinLobby(lobby.id)}>JOIN</button>
                                        </td>
                                    </tr>
                                ))
                            }
                        </tbody>
                    </table>
                </div>
                <div className='pagination'>
                    <button disabled={page === 0} onClick={() => setPage(page - 1)} className='btn btn-sm'>{'<'}</button>
                    <div className='page'>{page + 1}</div>
                    <button disabled={page === getLastPage()} onClick={() => setPage(page + 1)} className='btn btn-sm'>{'>'}</button>
                </div>
                <button className="btn w-100" onClick={() => close()}>
                    CLOSE
                </button>
            </div>
        </div>
    )
}
