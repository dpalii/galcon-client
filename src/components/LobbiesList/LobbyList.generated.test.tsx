import React from 'react';

import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { LobbyList } from '../../../src/components/LobbiesList/LobbyList';

let container: any = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('<LobbyList>', () => {
  it('should render component', () => {
    act(() => {
      render(<LobbyList
        joinLobby={(gameId: string) => {}}
        close={() => {}}
      />, container);
    });
    container.querySelector('.backdrop').dispatchEvent(new MouseEvent('click', { bubbles: true }));
    container.querySelector('.btn').dispatchEvent(new MouseEvent('click', { bubbles: true }));
    expect(container).toMatchSnapshot();
  });
});
