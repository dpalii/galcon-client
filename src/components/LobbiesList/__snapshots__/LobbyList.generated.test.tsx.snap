// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LobbyList> should render component 1`] = `
<div>
  <div
    class="backdrop"
  >
    <div
      class="modal"
    >
      <h2
        class="heading"
      >
        LOBBIES:
      </h2>
      <div
        class="lobbies-wrapper"
      >
        <table
          class="lobbies"
        >
          <thead>
            <tr>
              <th>
                ID
              </th>
              <th>
                User Amount
              </th>
              <th />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      <div
        class="pagination"
      >
        <button
          class="btn btn-sm"
          disabled=""
        >
          &lt;
        </button>
        <div
          class="page"
        >
          1
        </div>
        <button
          class="btn btn-sm"
          disabled=""
        >
          &gt;
        </button>
      </div>
      <button
        class="btn w-100"
      >
        CLOSE
      </button>
    </div>
  </div>
</div>
`;
